File: src/osgODE/Space.cpp
/*!
 * @file Space.cpp
 * @author Rocco Martino
 */
/***************************************************************************
 *   Copyright (C) 2010 - 2012 by Rocco Martino                            *
 *   martinorocco@gmail.com                                                *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Lesser General Public License as        *
 *   published by the Free Software Foundation; either version 2.1 of the  *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU Lesser General Public License for more details.                   *
 *                                                                         *
 *   You should have received a copy of the GNU Lesser General Public      *
 *   License along with this program; if not, write to the                 *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

/* ======================================================================= */
/* ....................................................................... */
#include <osgODE/Space>
#include <osgODE/RigidBody>
#include <osgODE/OverlappingPair>
#include <osgODE/DefaultNearCallback>
#include <osgODE/Notify>
#include <osgODE/ScopedTimer>
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
/* ....................................................................... */
/* ======================================================================= */




using namespace osgODE ;




/* ======================================================================= */
/* ....................................................................... */
static void
_defaultOverlappingPairsDispatcher(void* void_near_callback, dGeomID o1, dGeomID o2)
{

    Collidable* collidable1 = static_cast<Collidable*>(dGeomGetData(o1)) ;
    Collidable* collidable2 = static_cast<Collidable*>(dGeomGetData(o2)) ;



    osg::ref_ptr<OverlappingPair>   overlapping_pair = new OverlappingPair() ;

    overlapping_pair->setCollidables(collidable1, collidable2) ;



    PS_ASSERT1( void_near_callback != NULL ) ;


    //! @note   devo evitare questo static_cast da void*

    NearCallback*   near_callback = static_cast<NearCallback*>(void_near_callback) ;

    if( near_callback->collide(overlapping_pair) > 0 ) {

        // OK, a collision occours



        // call the collision callbacks

        CollisionCallback*  collision_callback1 = collidable1->getCollisionCallback() ;
        CollisionCallback*  collision_callback2 = collidable2->getCollisionCallback() ;


        if(collision_callback1) {
            (*collision_callback1)(collidable1, collidable2, near_callback) ;
        }


        if(collision_callback2) {
            (*collision_callback2)(collidable2, collidable1, near_callback) ;
        }
    }
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
Space::Space(void):
    m_overlapping_pairs_dispatcher(_defaultOverlappingPairsDispatcher),
    m_near_callback( new DefaultNearCallback() )
{

    m_ODE_space = dHashSpaceCreate(NULL) ;

    // The osgODE::Space will remove the bodies by itself
    dSpaceSetCleanup(m_ODE_space, 0) ;

    dGeomSetData((dGeomID)m_ODE_space, this) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
Space::Space(const Space& other, const osg::CopyOp& copyop):
    World(other, copyop),
    m_overlapping_pairs_dispatcher(other.m_overlapping_pairs_dispatcher),
    m_near_callback(other.m_near_callback.get())
{

    m_ODE_space = dHashSpaceCreate(NULL) ;

    _cloneODESpace(other.m_ODE_space, m_ODE_space) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
Space::~Space(void)
{
    removeAllObjects() ;

    dSpaceDestroy(m_ODE_space) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
Space::update(double step_size)
{
    collide() ;

    this->World::update( step_size ) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
Space::collide(void)
{
    PS_DBG3("osgODE::Space::collide(%p)", this) ;




    if( m_near_callback.valid() ) {

        m_near_callback->reset() ;

        {
            PS_DBG3("osgODE::Space::collide(%p): dSpaceCollide", this) ;
            PS_SCOPED_TIMER("dSpaceCollide") ;

            dSpaceCollide(m_ODE_space, (void*)m_near_callback, m_overlapping_pairs_dispatcher) ;
        }
    }
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
Space::copySpace(const Space* other)
{
    this->copyWorld(other) ;


    m_overlapping_pairs_dispatcher = other->m_overlapping_pairs_dispatcher ;

    m_near_callback = other->m_near_callback.get() ;


    _cloneODESpace(other->m_ODE_space, m_ODE_space) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
Space::_cloneODESpace(dSpaceID src, dSpaceID dst)
{
    PS_DBG2("osgODE::Space::_cloneODESpace(src=%p, dst=%p)", src, dst) ;

    dSpaceSetCleanup(dst, dSpaceGetCleanup(src)) ;
    dSpaceSetSublevel(dst, dSpaceGetSublevel(src)) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
bool
Space::addToWorldInternal(World* world)
{
    PS_DBG2("osgODE::Space::addToWorldInternal(%p, world=%p)", this, world) ;

    return true ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
bool
Space::removeFromWorldInternal(World* world)
{
    PS_DBG2("osgODE::Space::removeFromWorldInternal(%p, world=%p)", this, world) ;

    return true ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
Space*
Space::asSpace(void)
{
    return this ;
}
/* ....................................................................... */
/* ======================================================================= */

File: BlenderExporter/io_scene_sirengine/Space.py
# -*- coding: iso-8859-1 -*-
# file Space.py
# author Rocco Martino
#
############################################################################
#    Copyright (C) 2012 by Rocco Martino                                   #
#    martinorocco@gmail.com                                                #
#                                                                          #
#    This program is free software; you can redistribute it and/or modify  #
#    it under the terms of the GNU General Public License as published by  #
#    the Free Software Foundation; either version 2 of the License, or     #
#    (at your option) any later version.                                   #
#                                                                          #
#    This program is distributed in the hope that it will be useful,       #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of        #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
#    GNU General Public License for more details.                          #
#                                                                          #
#    You should have received a copy of the GNU General Public License     #
#    along with this program; if not, write to the                         #
#    Free Software Foundation, Inc.,                                       #
#    59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             #
############################################################################

############################################################################
from . import Writable, Game, NearCallback, Joint

import bpy
############################################################################




############################################################################
# ........................................................................ #
# ........................................................................ #
############################################################################




############################################################################
# ........................................................................ #
class Space(Writable.Writable):
    """osgODE::Space"""





############################################################################
    NearCallback = None
    ODEObjects = []
    Gravity = -9.80665
    ID = 0
    ERP = 0.2
    CFM = 1.0e-5
    RigidBodies = []
############################################################################









############################################################################
    def __init__(self, data):
        super(Space, self).__init__(data)

        self.NearCallback = None
        self.ODEObjects = []
        self.Gravity = -9.80665
        self.ID = 0
        self.ERP = 0.2
        self.CFM = 1.0e-5
        self.RigidBodies = []
############################################################################




############################################################################
    def buildGraph(self):
        super(Space, self).buildGraph()

        self.ID = self.Data.ODEID.generate()

        self.Gravity = -self.Data.Context.scene.game_settings.physics_gravity

        self.ERP = 0.2
        self.CFM = 1.0e-5

        self.NearCallback = NearCallback.NearCallback(self.Data)
        self.addChild(self.NearCallback)




        rigid_bodies = []
        empties = []


        for obj in self.Data.Scene.objects:
            if obj.is_visible( self.Data.Scene ) :
                if obj.type != "EMPTY" :
                    if not self.addObject(obj):
                        return False
                    else:
                        rigid_bodies.append(obj)
                else:
                    empties.append( obj )


        for obj in empties:
            if not self.addObject(obj):
                return False
            else:
                rigid_bodies.append(obj)



        for rigid_body in rigid_bodies:
            constraints = rigid_body.constraints

            for constraint in constraints:
                joint = Joint.Joint(self.Data, constraint, rigid_body)
                self.addChild(joint)
                self.ODEObjects.append(joint)



        if not self.traverseBuild():
            return False


        if self.Data.ExportGame:

            for rigid_body in self.RigidBodies:

                if ( len(rigid_body.Object.game.sensors) + len(rigid_body.Object.game.properties) ) > 0:

                    game = Game.Game( rigid_body )
                    game.buildGraph()

                    rigid_body.addUpdateCallback( game )


        return True
############################################################################




############################################################################
    def addObject(self, obj):

        if obj.type == 'MESH':

            num_materials = len(obj.material_slots)

            if num_materials > 1:
                self.Data.Operator.report({'ERROR'}, "[%s] This exporter supports only 1 material per mesh" %obj.name)
                return False



            is_graphics = False

            if num_materials == 0:
                is_graphics = True
            elif not obj.material_slots[0].material:
                is_graphics = True
            elif not obj.material_slots[0].material.game_settings.physics:
                is_graphics = True


            if not is_graphics:
                from . import ODEObject
                physics = self.selectODEObject(obj)

                if physics:
                    self.addChild(physics)
                    self.ODEObjects.append(physics)
                    self.RigidBodies.append(physics)


                if self.Data.Context.selected_objects and self.Data.Context.selected_objects[0] == obj:
                    self.Data.Selected = physics




        elif obj.type == 'EMPTY':

            from . import ODEObject
            physics = self.selectODEObject(obj)

            if physics:
                self.addChild(physics)
                self.ODEObjects.append(physics)
                self.RigidBodies.append(physics)


            if self.Data.Context.selected_objects and self.Data.Context.selected_objects[0] == obj:
                self.Data.Selected = physics



        return True
############################################################################




############################################################################
    def selectODEObject(self, obj):

        if obj.parent != None:
            return None


        elif not obj.game.use_collision_bounds:

            from . import RigidBody
            return RigidBody.RigidBody(self.Data, obj)



        elif obj.game.physics_type == 'NO_COLLISION' :

            from . import RigidBody
            return RigidBody.RigidBody(self.Data, obj)





        elif obj.game.physics_type in ['STATIC', 'DYNAMIC', 'RIGID_BODY']:
            if obj.game.collision_bounds_type == 'TRIANGLE_MESH':
                from . import TriMesh
                return TriMesh.TriMesh(self.Data, obj)

            elif obj.game.collision_bounds_type == 'BOX':
                from . import Box
                return Box.Box(self.Data, obj)

            elif obj.game.collision_bounds_type == 'SPHERE':
                from . import Sphere
                return Sphere.Sphere(self.Data, obj)

            elif obj.game.collision_bounds_type == 'CYLINDER':
                from . import Cylinder
                return Cylinder.Cylinder(self.Data, obj)

            elif obj.game.collision_bounds_type == 'CAPSULE':
                from . import Capsule
                return Capsule.Capsule(self.Data, obj)

            else:
                print("This exporter does not support %s collision bounds" %obj.game.collision_bounds_type)


            pass



        elif obj.game.physics_type in ['OCCLUDE']:
            # Esprtato in OccluderGroup
            return None



        else:
            print("This exporter does not support %s physics type" %obj.game.physics_type)
            return None
############################################################################




############################################################################
    def writeToStream(self, writer):
        writer.moveIn("osgODE::Space") ;

        if not super(Space, self).writeToStream(writer) :
            return False


        writer.writeLine("ID %d" %self.ID)

        writer.writeLine("Gravity 0 0 %f" %self.Gravity)

        writer.writeLine("ERP %f" %self.ERP)
        writer.writeLine("CFM %f" %self.CFM)


        writer.writeLine("WorldStepFunction %s" %self.Data.WorldStep)


        if len(self.ODEObjects):
            writer.writeLine("ODEObjects %d" %len(self.ODEObjects))

        for obj in self.ODEObjects:
            if not obj.writeToStream(writer):
                return False


        if self.Data.ExportGame:
            writer.moveIn("Events TRUE")
            writer.moveIn("osgODE::Events")
            writer.writeLine("UniqueID %d" % self.Data.UniqueID.generate() )
            writer.moveOut("osgODE::Events")
            writer.moveOut("Events TRUE")


        writer.moveIn("NearCallback TRUE")

        if not self.NearCallback.writeToStream(writer):
            return False

        writer.moveOut("NearCallback TRUE")



        writer.moveOut("osgODE::Space")

        return True
############################################################################




# ........................................................................ #
############################################################################
